src/ctapipe/monitoring/calculator.py

```diff
@@ -0,0 +1,337 @@
+            outlier_mask = np.logical_or(
+                outlier_mask,
+                outlier_detector(aggregated_stats[aggregated_val]),
+            )
+        # Add the outlier mask to the aggregated statistics
+        aggregated_stats["outlier_mask"] = outlier_mask
+        # Get valid chunks and add them to the aggregated statistics
+        aggregated_stats["is_valid"] = self._get_valid_chunks(outlier_mask)
+        return aggregated_stats
+
+    def second_pass(
+        self,
+        table,
+        valid_chunks,
+        tel_id,
+        masked_pixels_of_sample=None,
+        col_name="image",
+    ) -> Table:
+        """
+        Conduct a second pass over the data to refine the statistics in regions with a high percentage of faulty pixels.
+
+        This method performs a second pass over the data with a refined shift of the chunk in regions where a high percentage
+        of faulty pixels were detected during the first pass. Note: Multiple first passes of different calibration events are
+        performed which may lead to different identification of faulty chunks in rare cases. Therefore a joined list of faulty
+        chunks is recommended to be passed to the second pass(es) if those different passes use the same ``chunk_size``.
+
+        Parameters
+        ----------
+        table : astropy.table.Table
+            DL1-like table with images of shape (n_images, n_channels, n_pix), event IDs and timestamps of shape (n_images, ).
+        valid_chunks : ndarray
+            Boolean array indicating the validity of each chunk from the first pass.
+            Note: This boolean array can be a ``logical_and`` from multiple first passes of different calibration events.
+        tel_id : int
+            Telescope ID for which the calibration is being performed.
+        masked_pixels_of_sample : ndarray, optional
+            Boolean array of masked pixels of shape (n_pix, ) that are not available for processing.
+        col_name : str
+            Column name in the table from which the statistics will be aggregated.
+
+        Returns
+        -------
+        astropy.table.Table
+            Table containing the aggregated statistics after the second pass, their outlier masks, and the validity of the chunks.
+        """
+        # Check if the chunk_shift is set for the second pass
+        if self.chunk_shift is None:
+            raise ValueError(
+                "chunk_shift must be set if second pass over the data is requested"
+            )
+        # Get the aggregator
+        aggregator = self.stats_aggregator[self.stats_aggregator_type.tel[tel_id]]
+        # Conduct a second pass over the data
+        aggregated_stats_secondpass = None
+        faulty_chunks_indices = np.where(~valid_chunks)[0]
+        for index in faulty_chunks_indices:
+            # Log information of the faulty chunks
+            self.log.warning(
+                f"Faulty chunk detected in the first pass at index '{index}'."
```

Comment:
Don't use f-strings in the log messages. Weird this passed the CI, [G004 should be raised ](https://docs.astral.sh/ruff/rules/logging-f-string/)

@maxnoe FYI